feature-future-reservations: Add basic auto-generated API and CRD (work in progress)

fizashaikh opened the pull request:
### Change description

<!--
Describe what this pull request does.

* If your pull request is to address an open issue, indicate it by specifying the
issue number: 

For example: "Fixes #858"
-->
Fixes # https://github.com/GoogleCloudPlatform/k8s-config-connector/issues/4710

### Tests you have done

<!--

Make sure you have run "make ready-pr" to run required tests and ensure this PR is ready to review. 

Also if possible, share a bit more on the tests you have done. 

For example if you have updated the pubsubtopic sample, you can share the test logs from running the test case locally.

go test -v -tags=integration ./config/tests/samples/create -test.run TestAll -run-tests pubsubtopic

-->

- [ ] Run `make ready-pr` to ensure this PR is ready for review.
- [ ] Perform necessary E2E testing for changed resources.

fizashaikh commented on mockgcp/Makefile:
```diff
@@ -34,7 +34,7 @@ gen-proto: tools sync-repo
 .PHONY: gen-proto-no-fixup
 gen-proto-no-fixup:
-	PATH=bin/:${PATH} bin/protoc/bin/protoc \
+	PATH="bin/:${PATH}" bin/protoc/bin/protoc \
```
This prevents issues with spaces in PATH.

fizashaikh commented on scripts/generate-google3-docs/resource-reference/main.go:
```diff
@@ -181,7 +181,7 @@ func main() {
 		}
 		// TODO: Add resource docs for all the v1beta1 resources and remove exceptions.
-		if gvk.Kind == "KMSImportJob" || gvk.Kind == "MetastoreBackup" {
+		if gvk.Kind == "KMSImportJob" || gvk.Kind == "MetastoreBackup" || gvk.Kind == "BigQueryReservationReservation" {
```
Need to exclude `BigQueryReservationReservation` here to resolve error:
```
2025/08/15 19:21:28 error generating doc for GVK
bigqueryreservation.cnrm.cloud.google.com/v1beta1, 
Kind=BigQueryReservationReservation: error creating template: error 
parsing template files: open ~/Documents/Fiza/MyGithub/k8s-config-
connector/scripts/generate-google3-docs/resource-
reference/templates/bigqueryreservation_bigqueryreservationreservat
ion.tmpl: no such file or directory
```

fizashaikh commented:
/assign @justinsb

gemmahou commented on apis/compute/v1beta1/futurereservation_types.go:
```diff
@@ -0,0 +1,227 @@
+	ProjectRef *refv1beta1.ProjectRef `json:"projectRef"`
+
+	// +required
+	Location string `json:"location"`
```
Let's use "Zone" instead of "Location". Per [Compute documentation](https://cloud.google.com/compute/docs/regions-zones#identifying_a_region_or_zone) they are different.

gemmahou commented on docs/releasenotes/release-1.134.md:
```diff
@@ -0,0 +1,20 @@
+# Follow `SAMPLE_XXX` format to write the content.Add commentMore actions
```
Let's move this to release-1.135

fizashaikh replied:
Done, updated.

gemmahou replied:
Thanks for helping with the release notes! Our team has improved our release process, and the notes will be generated upon release, so we don't need to manually update them anymore. I will assign a correct milestone(release version) to this PR once it's ready to merge and make sure it's included in our release notes.

gemmahou commented on mockgcp/mockcompute/futurereservationsv1beta.go:
```diff
@@ -0,0 +1,216 @@
+}
+
+// newZonalLRO creates a zonal LRO for v1beta operation
+func (s *FutureReservationsV1beta) newZonalLRO(ctx context.Context, projectID string, zone string, op *pbv1beta.Operation) (*pbv1beta.Operation, error) {
```
Can we make this a shared function in operations.go, similar to "StartRegionalLRO"?

fizashaikh replied:
Done, updated.

gemmahou commented on mockgcp/mockcompute/zonaloperationsv1beta.go:
```diff
@@ -0,0 +1,74 @@
+	pbv1beta.UnimplementedZoneOperationsServer
+}
+
+func (s *ZoneOperationsV1Beta) Get(ctx context.Context, req *pbv1beta.GetZoneOperationRequest) (*pbv1beta.Operation, error) {
```
Wondering the necessity of the functions, I feel LROs are handled by startLRO0...

fizashaikh replied:
Hey @gemmahou, not sure if I understand correctly but the `startLRO0()` function accepts and returns a v1.Operation object whereas for Future Reservations I need v1beta1.Operation. Hence I have added a separate function for it. Does that answer your question?

gemmahou commented on pkg/controller/direct/compute/futurereservation_controller.go:
```diff
@@ -0,0 +1,300 @@
+	}
+	resource.Name = direct.LazyPtr(a.id.ID())
+
+	// Preserve API output-only fields
```
I'm unsure if we need this - CompareProtoMessage function should only compare spec fields and skip output-only fields.

fizashaikh replied:
Yes, I thought so but for some reason it cannot determine output only fields. Without settings these I keep getting,
```
{"severity":"info","timestamp":"2025-09-02T13:40:57.381+1200","msg":"resource is not 
ready","kind":"ComputeFutureReservation","name":"computefuturereservation-j7ita4gs6iqctoq","conditions":
[{"lastTransitionTime":"2025-09-02T13:40:09+12:00","message":"Update call failed: error updating: updating 
FutureReservation projects/wd-scyllasre-dev-1/zones/us-central1-a/futureReservations/computefuturereservation-
j7ita4gs6iqctoq: googleapi: Error 400: Invalid value for field 'updateMask': 
'\"creationTimestamp,id,kind,selfLink,selfLinkWithId,specificSkuProperties.totalCount,status.existingM...'. Invalid 
updateMask Path. Updating creation_timestamp is not 
supported.","reason":"UpdateFailed","status":"False","type":"Ready"}]}
```

gemmahou commented on pkg/controller/direct/compute/futurereservation_controller.go:
```diff
@@ -0,0 +1,300 @@
+		return err
+	}
+
+	// fields that cause drift or GCP API errors
```
Can we also add comments to the field description of these special fields?

fizashaikh replied:
Done, updated.

gemmahou commented on pkg/controller/direct/compute/futurereservation_controller.go:
```diff
@@ -0,0 +1,300 @@
+	}
+	op, err := a.gcpClient.Delete(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			// Return success if not found (assume it was already deleted).
+			log.V(2).Info("skipping delete for non-existent FutureReservation, assuming it was already deleted", "name", a.id)
+			return true, nil
+		}
```
Line 273-277 are no longer needed, as we handle non-existence in directbase-controller([code](https://github.com/GoogleCloudPlatform/k8s-config-connector/blame/ab4cf1cd160bd9a0a216b4b6c0bcfc9ce6d1e992/pkg/controller/direct/directbase/directbase_controller.go#L399-L402))

fizashaikh replied:
Done, removed it.

gemmahou commented on pkg/test/resourcefixture/testdata/basic/compute/v1beta1/computefuturereservation/create.yaml:
```diff
@@ -0,0 +1,31 @@
+# Copyright 2022 Google LLC
```
nit: update copyright year to 2025

fizashaikh replied:
Done, updated.

gemmahou commented on pkg/test/resourcefixture/testdata/basic/compute/v1beta1/computefuturereservation/create.yaml:
```diff
@@ -0,0 +1,31 @@
+# Copyright 2022 Google LLC
```
Could you also run the test against realGCP to collect real logs? We want to compare mock logs with the real to ensure its fidelity

fizashaikh replied:
Hey @gemmahou, yes I have generated the [golden object](https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/4801/files#diff-67e8602d66bd651948e8071f7d3cbe47d11c66782797da8d264f633227eb10d0) and [http log](https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/4801/files#diff-68dae583453263ab09ac5549598d717a5b028557a9e8c6edf59f08bdbf92c5d9) files by running against a real GCP env as mentioned [here](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/mockgcp/README.md#capture-golden-object-and-http-golden-logs).
Did you mean you want the output of whats printed on the stdout?

gemmahou reviewed (commented):
Thanks for your contribution!! I finished my initial review.

Feel free to lmk if there're any questions/concerns,

fizashaikh commented on apis/compute/v1beta1/futurereservation_identity.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+func (i *FutureReservationIdentity) String() string {
+	return i.parent.String() + "/futurereservations/" + i.id
```
done

fizashaikh commented on apis/compute/v1beta1/futurereservation_identity.go:
```diff
@@ -0,0 +1,117 @@
+
+func ParseFutureReservationExternal(external string) (parent *FutureReservationParent, resourceID string, err error) {
+	tokens := strings.Split(external, "/")
+	if len(tokens) != 6 || tokens[0] != "projects" || tokens[2] != "locations" || tokens[4] != "futurereservations" {
```
Done, updated.

fizashaikh commented on apis/compute/v1beta1/futurereservation_reference.go:
```diff
@@ -0,0 +1,83 @@
+// holds the GCP identifier for the KRM object.
+type FutureReservationRef struct {
+	// A reference to an externally managed ComputeFutureReservation resource.
+	// Should be in the format "projects/{{projectID}}/locations/{{location}}/futurereservations/{{futurereservationID}}".
```
Done, updated.

fizashaikh commented:
> Thanks for your contribution!! I finished my initial review.
> 
> Feel free to lmk if there're any questions/concerns,

Hey @gemmahou, thank you for taking the time to review the PR! :-)
I have addressed most of the comments but also left questions (as replies to each comment) for a couple of them which I did not quite understand. 
Could you please check? 
Thanks in advance!

fizashaikh commented:
> > > Thanks for your contribution!! I finished my initial review.
> > > Feel free to lmk if there're any questions/concerns,
> > 
> > 
> > Hey @gemmahou, thank you for taking the time to review the PR! :-) I have addressed most of the comments but also left questions (as replies to each comment) for a couple of them which I did not quite understand. Could you please check? Thanks in advance!
> 
> Thanks @fizashaikh! Sorry for the late reply, are you still responsible for this PR? It has been several months since my last review, and there're some of merge conflicts need to be resolved. I'll perform a re-review and address your questions.

Hi @gemmahou, yes I am.
I had to discontinue working on this PR since I was told that future reservations are going into GA soon so I should wait. Since this PR was created using v1beta1 api, I think after GA that will have to change? Or is it still okay? 
If that is so, I can work on resolving conflicts. 
If not, I may need more time to move away from the v1beta1 API. I might as well raise a fresh PR in future in that case since that would be more simpler.
Let me know your thoughts. Thanks in advance!

fkc1e100 commented:
This PR addresses feature request issue #6828

gemmahou commented:
> > > > Thanks for your contribution!! I finished my initial review.
> > > > Feel free to lmk if there're any questions/concerns,
> > > 
> > > 
> > > Hey @gemmahou, thank you for taking the time to review the PR! :-) I have addressed most of the comments but also left questions (as replies to each comment) for a couple of them which I did not quite understand. Could you please check? Thanks in advance!
> > 
> > 
> > Thanks @fizashaikh! Sorry for the late reply, are you still responsible for this PR? It has been several months since my last review, and there're some of merge conflicts need to be resolved. I'll perform a re-review and address your questions.
> 
> Hi @gemmahou, yes I am. I had to discontinue working on this PR since I was told that future reservations are going into GA soon so I should wait. Since this PR was created using v1beta1 api, I think after GA that will have to change? Or is it still okay? If that is so, I can work on resolving conflicts. If not, I may need more time to move away from the v1beta1 API. I might as well raise a fresh PR in future in that case since that would be more simpler. Let me know your thoughts. Thanks in advance!

Thanks for the quick response! Our team is currently leveraging @codebot-robot to help us accelerate the dev process(https://github.com/GoogleCloudPlatform/k8s-config-connector/tree/master/docs/develop-resources#1-ai-assisted-workflow-recommended). Frank and I have linked this PR to the open issue associated with this feature request, and we will monitor its progress.

fizashaikh commented:
> > > > > Thanks for your contribution!! I finished my initial review.
> > > > > Feel free to lmk if there're any questions/concerns,
> > > > 
> > > > 
> > > > Hey @gemmahou, thank you for taking the time to review the PR! :-) I have addressed most of the comments but also left questions (as replies to each comment) for a couple of them which I did not quite understand. Could you please check? Thanks in advance!
> > > 
> > > 
> > > Thanks @fizashaikh! Sorry for the late reply, are you still responsible for this PR? It has been several months since my last review, and there're some of merge conflicts need to be resolved. I'll perform a re-review and address your questions.
> > 
> > 
> > Hi @gemmahou, yes I am. I had to discontinue working on this PR since I was told that future reservations are going into GA soon so I should wait. Since this PR was created using v1beta1 api, I think after GA that will have to change? Or is it still okay? If that is so, I can work on resolving conflicts. If not, I may need more time to move away from the v1beta1 API. I might as well raise a fresh PR in future in that case since that would be more simpler. Let me know your thoughts. Thanks in advance!
> 
> Thanks for the quick response! Our team is currently leveraging @codebot-robot to help us accelerate the dev process(https://github.com/GoogleCloudPlatform/k8s-config-connector/tree/master/docs/develop-resources#1-ai-assisted-workflow-recommended). Frank and I have linked this PR to the open issue associated with this feature request, and we will monitor its progress.

Thanks @gemmahou! We are really looking forward to have this feature. 
Let me know if you need anything from me :-)
